src/Data/Array/NonEmpty.purs

```diff
@@ -166,12 +168,22 @@ unsafeFromArray = NonEmptyArray
 unsafeFromArrayF :: forall f a. f (Array a) -> f (NonEmptyArray a)
 unsafeFromArrayF = unsafeCoerce
 
+-- Note that this is unsafe: if the array or any embedded array is empty, this can
+-- explode at runtime.
+unsafeFromArray2D :: forall a. Array (Array a) -> NonEmptyArray (NonEmptyArray a) 
+unsafeFromArray2D =  
+  (map unsafeFromArray) <<< unsafeFromArray
+
 fromNonEmpty :: forall a. NonEmpty Array a -> NonEmptyArray a
 fromNonEmpty (x :| xs) = cons' x xs
 
 toArray :: forall a. NonEmptyArray a -> Array a
 toArray (NonEmptyArray xs) = xs
 
+toArray2D :: forall a. NonEmptyArray (NonEmptyArray a) -> Array (Array a)
+toArray2D =
+  (map toArray) <<< toArray
```

Comment:
I believe you should be using `coerce` here (see https://pursuit.purescript.org/packages/purescript-safe-coerce/2.0.0/docs/Safe.Coerce#v:coerce). Otherwise, you're literally creating a copy of the original array with no change in the elements due to the usage of `map`. That's the very problem `coerce` solves.

That will require adding a dependency to `purescript-safe-coerce`.

Reply:
That's a good note, but might beg the question of a larger change, since the current `unsafeAdapt`, and other related coercions would also probably benefit from `coerce`.

If it wasn't totally clear from Jordan's suggestion - the reason `coerce` is better than this suggested code is it can "deep coerce" without the need for mapping and such, hence no copying of array.

We could also just shoehorn an `unsafeCoerce` in, since that's already used in here. Would be nice to use `coerce` where we can though.

Reply:
> We could also just shoehorn an `unsafeCoerce` in, since that's already used in here.

I disagree. `unsafeCoerce` made sense when we didn't have `Coercible` support. Now that we have support for it, let's design the code right.

Reply:
So does this mean that both functions in question can just be implemented simple with ```coerce```   and that this is somehow safe so that ```unsafeFromArray2d``` could now be called ```fromArray2D``` ?

Reply:
Yes and no :smile:

Both should be a single `coerce`, but there are two independent "safeties" involved here:

1. That he type being coerced is represented as an array. That's guaranteed by using `coerce`.
2. Whether the invariant that the array is actually non-empty is preserved, and we're not tracking that in the type system.

So going **to** `NonEmptyArray` is completely safe using `coerce`, going **from** `Array` we either need to check the invariants and make the function return `Maybe`, or mark it as unsafe, because it's up to the caller to have checked for non-emptiness first.

Reply:
Yes I thought so - presumably ```coerce``` would have to rely on the safety of the ```NonEmptyArray``` constructor and this is manifestly unsafe.  I think I'm beginning to understand.  I hadn't realised before that coerce was deep, and so never thought that using it was a possibility.

Reply:
I've made the change to use ```coerce```